### ISSUE TITLE
Investigate ways to speed up CI/CD for front end

### ISSUE THREAD

**Issue by vidjovanovic** [department-of-veterans-affairs/va.gov-team#69650]
Per Josh: Reach out to Alex when ready to take on this ticket.

**Comment by vidjovanovic** [comment]
Make slack discussion group with Alex and rest of DevOps

**Comment by asg5704** [comment]
This is an example of CI/CD on the FE:
https://github.com/department-of-veterans-affairs/vets-website/actions/runs/7462295957?pr=27396

So a couple of things to note:
- The total duration is more than 1 hour
    - Ideally I'd like CI/CD for FE to be no longer than 15-20 minutes if possible
- When failures occur during a job run in CI/CD there is no fast-failure, other jobs keep running
    - I'd like for any Required label jobs (Unit Tests, E2E Tests, Builds, etc) to have a fast-failure/early-fail feature and skip all remaining jobs
- Annotations list
    - There are several warnings on the `set-output` command will be disabled soon that I would like fixed
- Other performance features if available
    - Basically all I want to be able to do is put up a PR and have CI/CD go as fast as possible for the Frontend and however that can be accomplished will be amazing.
    - Recommendations from a DevOps perspective on how vets-website CI/CD can be improved

CI/CD times a variable
- https://github.com/department-of-veterans-affairs/vets-website/actions/runs/7462295957?pr=27396 (took 1 hour of CI/CD time)
- https://github.com/department-of-veterans-affairs/vets-website/actions/runs/7462481748?pr=27397 (took 20 minutes of CI/CD time, with a failure)
- https://github.com/department-of-veterans-affairs/vets-website/actions/runs/7425586314?pr=27364 (took 42 minutes of CI/CD time, with a success)

**Comment by asg5704** [comment]
@vidjovanovic @jhouston0 see above. Sorry I thought I tagged you all previously

**Comment by jhouston0** [comment]
@asg5704 just wanted to update you on the ticket, we have a gameplan going and moving on to the testing phase for testing the speed optimizations before we present it to you

**Comment by jhouston0** [comment]
Ok, so we have done a fair amount of testing; the findings are below

## Build Container steps:
1. Available runners. The biggest issue for a fast build time is there are just not enough runners. When testing the runners, on average, it took 10-15 minutes for the job to get picked up. This issue affects not just the build steps but throughout the build pipeline 
 <img width="1423" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBL1RUQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f0128326414eaa37134a7c797983d6e814492c7c/Screenshot%202024-03-01%20at%2012.22.01%E2%80%AFPM.png" alt="Screenshot 2024-03-01 at 12.22.01 PM.png" />
 <img width="1335" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNS9UQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--db8ca43812a539f39d494f98ad0883e06b87c9d6/Screenshot%202024-02-13%20at%203.12.25%E2%80%AFPM.png" alt="Screenshot 2024-02-13 at 3.12.25 PM.png" />

2. Tar and untaring the results. Compressing and archiving the build takes almost as long as building the application. If we separate the build step, make a compress and archive build step separately, and pass the application folders to other steps, that would knock down the build time by around 50%. We would also save time on all tests that have a uncompress feature
<img width="1494" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNmpUQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e89b11a48bd5b280993fa47b4fd1185001ce390e/image.png" alt="image.png" />
3. I tested  requesting more memory from the runners and modifying me memory in the build script, but no real speed difference was observed 
## Unit tests:
 A fair amount of testing went into this. Admittedly, many efforts and tests I ran were futile due to the broken unit tests. Once a fix was out, I worked on re-testing. As of when this ticket was written, all unit tests aren't working. Additional tickets will be created for specific subtasks of unit testing optimization. Here are some (not all) of the tests I ran:
 1. Changing the build:webpack space size FROM 8192 TO 16384 in package.json
 2. Changing the webpack args in build.sh to add the option `--env parallel=true`
 3. in run-unit-test.js, changing the max workers to 50%
 3. in run-unit-test.js, changing the space size from 4096 to 8192
 4. re-writing mocha.js to use mocha parallel

## Fail fast on failure requests: 
There's a setting in the GitHub workflows:
 `strategy:
      fail-fast: false`
Which accomplishes this task. However, I would recommend keeping it off, as it doesn't do much with how our pipeline works. The reason being is most of the CI/CD time is spent running code, so a fail-fast feature on the CI/CD platform isn't worth it 

## Not reviewed:
I didn't review the following steps in the pipeline due to them already being pretty fast, they always complete before the build step is done:
  1. Fetch Test Stability
  2. Testing Reports Prep
  3. Security Audit
